src/manage_license.c

```diff
@@ -0,0 +1,225 @@
+
+  data->meta = license_meta_new ();
+  data->hardware = license_hardware_new ();
+
+  data->features = g_tree_new_full ((GCompareDataFunc) g_ascii_strcasecmp,
+                                    NULL, g_free, NULL);
+
+  data->limits = g_tree_new_full ((GCompareDataFunc) g_ascii_strcasecmp,
+                                  NULL, g_free, NULL);
+
+  data->keys = g_tree_new_full ((GCompareDataFunc) g_ascii_strcasecmp,
+                                NULL, g_free, g_free);
+
+  data->signature = g_tree_new_full ((GCompareDataFunc) g_ascii_strcasecmp,
+                                     NULL, g_free, g_free);
+
+  return data;
+}
+
+/**
+ * @brief Frees a license data struct and its fields.
+ *
+ * @param[in]  data  The data struct to free.
+ */
+void
+license_data_free (license_data_t *data)
+{
+  if (data == NULL)
+    return;
+
+  license_meta_free (data->meta);
+  license_hardware_free (data->hardware);
+  g_tree_destroy (data->features);
+  g_tree_destroy (data->limits);
+  g_tree_destroy (data->keys);
+  g_tree_destroy (data->signature);
+
+  g_free (data);
+}
+
+/* Actions */
+
+/**
+ * @brief Update the license file by replacing it with the given one.
+ *
+ * @param[in]  new_license  The content of the new license.
+ *
+ * @return 0 success, 99 permission denied.
+ */
+int
+manage_update_license_file (const char *new_license)
+{
+  if (! acl_user_may ("modify_license"))
+    return 99;
+
+  g_message ("%s: Uploaded new license file (%lu bytes)",
+             __func__, strlen (new_license));
+
+  return 0;
+}
+
+/**
+ * @brief Get the current license information.
+ *
+ * @param[out] status       The validation status (e.g. "valid", "expired").
+ * @param[out] license_data The content of the license organized in a struct.
+ * @param[out] file_content The TOML file content.
+ *
+ * @return 0 success, 1 service unavailable, 99 permission denied.
+ */
+int
+manage_get_license (gchar **status,
+                    license_data_t **license_data,
+                    gchar **file_content)
+{
+  if (! acl_user_may ("get_license"))
+    return 99;
+
+  if (status)
+    *status = g_strdup ("valid");
+
+  if (license_data)
+    {
+      *license_data = license_data_new ();
+      license_meta_t *license_meta = (*license_data)->meta;
+      license_hardware_t *license_hardware = (*license_data)->hardware;
+
+      // TODO : replace dummy data with data from license service
+      license_meta->id = g_strdup ("4711");
+      license_meta->schema_version = 1;
+      license_meta->title = g_strdup ("Test License");
+      license_meta->license_type = g_strdup ("Trial");
+      license_meta->customer_name = g_strdup ("Jane Doe");
+      license_meta->created = time (NULL) - 3600;
+      license_meta->begins = time (NULL);
+      license_meta->expires = time (NULL) + 3600 * 24 * 8;
+
+      license_hardware->model = g_strdup ("GSM XYZ");
+      license_hardware->model_type = g_strdup ("Virtual Appliance");
```

Comment:
Should be either "virtual" or "hardware"
```suggestion
      license_hardware->model_type = g_strdup ("virtual");
```